In [1]:
# 정수

a = 1000
print(a)

a = -7
print(a)

a = 0
print(a)

1000
-7
0


In [2]:
# 실수

# 양의 실수
a = 157.12
print(a)


# 음의 실수
a = -157.12
print(a)

# 정수부가 0일 때 0을 생략
a = -.7
print(a) # -0.7

157.12
-157.12
-0.7


In [3]:
# 10억의 지수 표현 방식

a = 1e9
print(a)

# 752.5
a = 75.25e1
print(a)

# 3.954
a = 3954e-3
print(a)

1000000000.0
752.5
3.954


In [6]:
# 실수 표현에 대한 한계

a = 0.3 + 0.6
print(a)

if a == 0.9:
  print(True)
else :
  print(False)

# round 함수 사용해보기
# round(실수형, 소수점 이하 몇 자리까지 남길지(반올림 위치))
print(round(123.456, 2))

a = 0.3 + 0.6
print(round(a, 4))

if round(a, 4) == 0.9:
  print(True)
else :
  print(False)

0.8999999999999999
False
123.46
0.9
True


`print(round(a, 4))`
- a = 0.8999999999999999
- round(a, 4) → 소수점 넷째 자리까지 반올림 → "0.9" 출력됨

In [9]:
a = 7
b = 3

# 자료형 연산

print(a+b)
print(a-b)
print(a*b)

# / 나눗셈은 실수형으로 나온다
print(a/b)

# 나머지
print(a%b)

# 몫
print(a//b)

# 거듭제곱
print(a**2)

10
4
21
2.3333333333333335
1
2
49


In [13]:
# 리스트 자료형

a = [1,2,3,4,5,6,7,8,9]
print(a)

# 인덱싱 (4번째 인덱스 즉, 다섯 번째 원소에 접근)
print(a[4]) # 5

# 빈 리스트 선언 방법_1
a_list = list()
print(a_list)

# 빈 리스트 선언 방법_2
a_list2 = []
print(a_list2)

[1, 2, 3, 4, 5, 6, 7, 8, 9]
5
[]
[]


In [14]:
# 크기가 n이고, 모든 값이 0인 1차원 리스트 초기화

n = 10
a = [0] * n
print(a)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


`a = [0] * n`
- a = [0] : [0]은 원소 하나(0)만 있는 리스트
- `* n`은 리스트 반복 연산자
- 즉, 길이가 10이고 모든 원소가 0인 리스트를 생성.

In [18]:
# 리스트 인덱싱

a = [1,2,3,4,5,6,7,8,9]
print(a[-1])

print(a[-3])

a[3] = 7
print(a)

# 슬라이싱
print(a[1:4]) # 2,3,7
print(a[::2]) # 1,3,5,7,9

9
7
[1, 2, 3, 7, 5, 6, 7, 8, 9]
[2, 3, 7]
[1, 3, 5, 7, 9]


In [19]:
# 리스트 컴프리헨션

# 0부터 19까지의 수 중에서 홀수만 포함하는 리스트

array = [i for i in range(20) if i % 2 == 1]

print(array)

'''
같은 코드

array = []

for i in range(20):
  if i % 2 == 1:
    array.append(i)

print(array)
'''

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]


`array = [i for i in range(20) if i % 2 == 1]`
- `range(20)` : 0부터 19까지 정수 생성 → [0, 1, 2, ..., 19]

- `for i in range(20)` : 0부터 19까지 하나씩 i에 대입하며 반복.

- `if i % 2 == 1` : i % 2는 i를 2로 나눈 나머지.== 1 → 홀수만 선택.

- `[i for i in range(20) if i % 2 == 1]` :

  - 리스트 내포(list comprehension) 문법.

  - range(20) 안의 홀수만 뽑아 리스트로 만듦.

  ---

## 컴프리헨션(comprehension)
- 파이썬에서 리스트, 집합, 딕셔너리를 간단하고 직관적으로 만드는 문법

### 📌 리스트 컴프리헨션 기본 구조
```python
[표현식 for 변수 in 반복가능객체 if 조건식]
```
- 표현식 → 리스트에 넣고 싶은 값
- for 변수 in 반복가능객체 → 반복문
- if 조건식 → (선택사항) 조건이 맞을 때만 추가

#### 예시
- 0에서 4까지 제곱수를 리스트에 넣고 싶을 때 ...

**원래 반복문**
```python
squares = []
for i in range(5):
    squares.append(i * i)

print(squares)  # [0, 1, 4, 9, 16]
```

**컴프리헨션**
```python
squares = [i * i for i in range(5)]
print(squares)  # [0, 1, 4, 9, 16]
```

In [20]:
# 1부터 9까지의 수의 제곱 값을 포함하는 리스트

array = [i * i for i in range(1,10)]

print(array)

[1, 4, 9, 16, 25, 36, 49, 64, 81]


In [22]:
# N X M 크기의 2차원 리스트 초기화

n = 3
m = 3

array = [[0] * m for _ in range(n)]
print(array)

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]


`[[0] * m for _ in range(n)]`

- `[0] * m → [0, 0, 0]` :  (길이 m인 리스트)

- `for _ in range(n)` → 위 리스트를 n번 반복해서 새로 생성

- 즉, 3×3 크기의 2차원 리스트가 만들어짐.

In [25]:
# 잘못된 초기화 방법

n = 3
m = 3
array = [[0] * m for _ in range(n)]
print(array)

array[1][1] = 5
print(array)


# 내부적으로 포함된 4개의 리스트 모두가 동일한 객체에 대한 3개의 레퍼런스로 인식하기 때문에
# 특정한 크기를 가지는 2차원 리스트를 초기화할 때는 컴프리헨션을 이용하면 좋다

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
[[0, 0, 0], [0, 5, 0], [0, 0, 0]]


In [34]:
# 리스트 관련 메소드

# 기본 리스트
a = [1,4,3]
print(f'기본 리스트 : {a}')

# 리스트에 맨 뒤에 원소 삽입
a.append(2)
print(f'사입 : {a}')

# 오름차순 정렬
a.sort()
print(f'오름차순 정렬 : {a}')

# 내림차순 정렬
a.sort(reverse=True)
print(f'내림차순 정렬 : {a}')

# 리스트 원소 뒤집기
a.reverse()
print(f'원소 뒤집기 : {a}')

# 특정 인덱스에 데이터 추가하기
a.insert(2,3)
print(f'인덱스 2에 3 추가 : {a}') # [1,2,3,3,4]

# 특정 값인 데이터 개수 세기
print(f'값이 3인 데이터 개수 : {a.count(3)} 개')

# 특정 값 데이터 삭제
a.remove(1)
print(f'값이 1인 데이터 삭제 : {a}')

기본 리스트 : [1, 4, 3]
사입 : [1, 4, 3, 2]
오름차순 정렬 : [1, 2, 3, 4]
내림차순 정렬 : [4, 3, 2, 1]
원소 뒤집기 : [1, 2, 3, 4]
인덱스 2에 3 추가 : [1, 2, 3, 3, 4]
값이 3인 데이터 개수 : 2 개
값이 1인 데이터 삭제 : [2, 3, 3, 4]


## 🕓 각 리스트 함수별 시간 복잡도

1. **append()**
- 시간 복잡도: *O(1)* (상수 시간)
- 리스트 맨 끝에 추가하므로 매우 빠릅니다
```python
pythonmy_list.append(x)  # O(1)
```
2. **insert()**
- 시간 복잡도: *O(n)* (선형 시간)
- 중간에 삽입하면 뒤의 모든 원소를 한 칸씩 이동해야 합니다
```python
pythonmy_list.insert(0, x)     # O(n) - 맨 앞 삽입이 최악
my_list.insert(len(my_list), x)  # O(1) - 맨 뒤는 append와 동일
```
3. **remove()**
- 시간 복잡도: *O(n)* (선형 시간)
- 값을 찾기 위해 리스트를 순회하고, 삭제 후 뒤의 원소들을 이동해야 합니다
```python
pythonmy_list.remove(x)  # O(n) - 값 찾기 + 원소 이동
```

In [35]:
a = [1,2,3,4,5,5,5,]
remove_set = {3,5}

# remove_set에 포함되지 않은 값만 저장
result = [i for i in a if i not in remove_set]
print(result)

[1, 2, 4]


In [38]:
# 문자열
data = 'Hello world'
print(data)

data = "Don't you know \"Python\"?"
print(data)

Hello world
Don't you know "Python"?


In [42]:
# 문자열 연산
a = "Hello"
b = 'World'
print(a+b)

a = 'String'
print(a * 3)

a = 'ABCDEF'
print(a[2:4]) # CD

HelloWorld
StringStringString
CD


In [44]:
# 튜플 자료형
# 한 번 선언된 값은 변경할 수 없다
# 소괄호 사용

a = (1,2,3,4)
print(a)

# a[2] = 7
# print(a)

(1, 2, 3, 4)


In [46]:
# 딕셔너리 자료형
# 키-값
# 순서 없음 : 인덱싱, 슬라이싱 못함

data = dict()

data['사과'] = 'apple'
data['바나나'] = 'banana'
data['코코넛'] = 'coconut'

print(data)

if '사과' in data:
  print('\'사과\'를 키로 가지는 데이터가 존재합니다.')

{'사과': 'apple', '바나나': 'banana', '코코넛': 'coconut'}
'사과'를 키로 가지는 데이터가 존재합니다.


In [48]:
# 딕셔너리 자료형 관련 함수

data = dict()
data['사과'] = 'apple'
data['바나나'] = 'banana'
data['코코넛'] = 'coconut'

# 키 데이터만 담은 리스트
key_list = data.keys()
print(key_list)
print()

# 값 데이터만 담은 리스트
value_list = data.values()
print(value_list)
print()

# 각 키에 따른 값을 하니씩 출력
for key in key_list:
  print(data[key])

dict_keys(['사과', '바나나', '코코넛'])

dict_values(['apple', 'banana', 'coconut'])

apple
banana
coconut


In [50]:
# 집합 자료형
# 중복을 허용하지 않는다
# 순서 없음 : 인덱싱, 슬라이싱 못함
# 값 데이터만 가지게 된다.
# 중괄호 사용

data = set([1,1,2,3,4,4,4])
print(data)

data = {1,1,2,3,4,4,4}
print(data)

{1, 2, 3, 4}
{1, 2, 3, 4}


In [52]:
a = set([1,2,3,4,5])
b = set([3,4,5,6,7])

print(f'합집합 : {a|b}')
print(f'교집합 : {a&b}')
print(f'차집합 : {a-b}')

합집합 : {1, 2, 3, 4, 5, 6, 7}
교집합 : {3, 4, 5}
차집합 : {1, 2}


In [53]:
# 집합 자료형 관련 함수

data = set([1,2,3])
print(data)

# 새로운 원소 추가
data.add(4)
print(data)

# 새로운 원소 여러 개 추가
data.update([5,6])
print(data)

# 특정한 값을 갖는 원소 삭제
data.remove(3)
print(data)

{1, 2, 3}
{1, 2, 3, 4}
{1, 2, 3, 4, 5, 6}
{1, 2, 4, 5, 6}
